scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala

```diff
@@ -251,9 +252,12 @@ class Router(formatOps: FormatOps) {
           case Newlines.keep if newlines != 0 => None
           case Newlines.unfold => None
           case Newlines.fold =>
+            val isTopLevelBlock =
+              leftOwner.parent.exists(_.parent.isEmpty) ||
+                (leftOwner.is[Template] &&
+                  !leftOwner.parent.exists(_.is[NewAnonymous]))
```

Comment:
i wanted to exclude from folding cases like:
```
  new A {
    def foo = bar
  }
```
i think there's a limit to what can be put on one line and still be readable, and a class definition seemed like a good cases to disallow folding.

however, your test example of `new Bar { () => f(42) }` looks like something else. perhaps we should check the `lambdaPolicy` first?

Reply:
Yes, I missed this case. Thank you, will rework it

Reply:
@poslegm i looked at this change trying to understand why it's different, and ended up changing a few things to make sure `squash` is applied in this case as well. (also reordered and added some tests, i like adding tests before a change and then see how it modifies their behaviour.)

perhaps you can use some of this:

https://github.com/kitbellew/scalafmt/commits/issue-1281

Reply:
Thank you for participating! Your implementation looks more reliable and consistent. I commited it here https://github.com/scalameta/scalafmt/pull/1833/commits/83a36d372039e17f8c8006140feb101856b491f0. The only problem is that it still folds constructors with `Defn`:

```
<<< 
 new A {
    def foo = bar
  }
>>>
new A { def foo = bar }
```

I changed it here: https://github.com/scalameta/scalafmt/pull/1833/commits/873363531ac1dc541f13c8ed0f4f7733f5e179a2